### ISSUE TITLE
Implementation of OPTRAM algorithm to derive soil moisture from remote sensing imagery 

### ISSUE THREAD

**Issue by micha-silver** [ropensci/software-review#612]
**Date accepted:** 2024-06-05

Submitting Author Name: Micha Silver
Submitting Author Github Handle: @micha-silver
Other Package Authors Github handles: (comma separated, delete if none) @github_handle1, @github_handle2
Repository:  https://gitlab.com/rsl-bidr/roptram
Version submitted: 0.0.1.000
Submission type: Standard
Editor: @adamhsparks
Reviewers: @harryeslick, @obrl-soil

Archive: TBD
Version accepted: TBD
Language: en

---

-   Paste the full DESCRIPTION file inside a code block below:

```
Package: rOPTRAM
Title: Derive soil moisture using the OPTRAM algorithm
Version: 0.0.1.000
Authors@R: c(
  person("Micha", "Silver", , "silverm@post.bgu.ac.il", role = c("aut", "cre"),
           comment = c(ORCID = "0000-0002-1128-1325")),
  person("Arnon", "Karnieli", , "karnieli@bgu.ac.il", role = c("ctb", "fnd"),
          comment = c(ORCID = "0000-0001-8065-9793"))
  )
Description: The OPtical TRapezoid Model (OPTRAM) derives soil moisture
  based on the linear relation between a vegetation index
  and Land Surface Temperature (LST). The Short Wave Infra-red (SWIR) band
  is used as a proxy for LST. See:
  Sadeghi, M., Babaeian, E., Tuller, M., Jones, S.B., 2017.
  The optical trapezoid model:
  A novel approach to remote sensing of soil moisture 
  applied to Sentinel-2 and Landsat-8 observations.
  Remote Sensing of Environment 198, 52–68,
  https://doi.org/10.1016/j.rse.2017.05.041 .
License: GPL (>= 3) + file LICENSE
URL: https://gitlab.com/rsl-bidr/roptram.git
BugReports: https://gitlab.com/rsl-bidr/
Encoding: UTF-8
Roxygen: list(markdown = TRUE)
RoxygenNote: 7.2.3
Depends:
  R (>= 4.1.0)
Imports:
    dplyr,
    ggplot2,
    sf,
    terra,
    tools,
    utils
Suggests:
    geojsonio,
    geojsonlint,
    stats,
    sen2r (> 1.5.0),
    testthat (>= 3.0.0),
    xml2
Config/testthat/edition: 3
```

## Scope

- Please indicate which category or categories from our [package fit policies](https://ropensci.github.io/dev_guide/policies.html#package-categories) this package falls under: (Please check an appropriate box below. If you are unsure, we suggest you make a pre-submission inquiry.):

	- [x] data retrieval
	- [x] data extraction
	- [ ] data munging
	- [ ] data deposition
    - [ ] data validation and testing
	- [ ] workflow automation
	- [ ] version control
	- [ ] citation management and bibliometrics
	- [ ] scientific software wrappers
	- [ ] field and lab reproducibility tools
	- [ ] database software bindings
	- [x] geospatial data
	- [ ] text analysis

- Explain how and why the package falls under these categories (briefly, 1-2 sentences):
This package includes acquiring satellite imagery, and preparing spatially explicit soil moisture raster grids.

-   Who is the target audience and what are scientific applications of this package?
Researchers in ecology, agriculture, sustainability. Agricultural management of grazing lands, reforestration.

-   Are there other R packages that accomplish the same thing? If so, how does yours differ or meet [our criteria for best-in-category](https://ropensci.github.io/dev_guide/policies.html#overlap)?
No

-   (If applicable) Does your package comply with our [guidance around _Ethics, Data Privacy and Human Subjects Research_](https://devguide.ropensci.org/policies.html#ethics-data-privacy-and-human-subjects-research)?
Yes

-   If you made a pre-submission inquiry, please paste the link to the corresponding issue, forum post, or other discussion, or @tag the editor you contacted.

-   Explain reasons for any [`pkgcheck` items](https://docs.ropensci.org/pkgcheck/) which your package is unable to pass.

## Technical checks

Confirm each of the following by checking the box.

- [x] I have read the [rOpenSci packaging guide](https://devguide.ropensci.org/building.html).
- [x] I have read the [author guide](https://devdevguide.netlify.app/authors-guide.html) and I expect to maintain this package for at least 2 years or to find a replacement.

This package:

- [x] does not violate the Terms of Service of any service it interacts with.
- [x] has a CRAN and OSI accepted license.
- [x] contains a [README with instructions for installing the development version](https://ropensci.github.io/dev_guide/building.html#readme).
- [x] includes [documentation with examples for all functions, created with roxygen2](https://ropensci.github.io/dev_guide/building.html#documentation).
- [x] contains a vignette with examples of its essential functions and uses.
- [x] has a [test suite](https://ropensci.github.io/dev_guide/building.html#testing).
- [x] has [continuous integration](https://ropensci.github.io/dev_guide/ci.html), including reporting of test coverage.

## Publication options

- [ ] Do you intend for this package to go on CRAN?
- [ ] Do you intend for this package to go on Bioconductor?

- [x] Do you wish to submit an Applications Article about your package to [Methods in Ecology and Evolution](http://besjournals.onlinelibrary.wiley.com/hub/journal/10.1111/(ISSN)2041-210X/)? If so:

<details>
<summary>MEE Options</summary>

- [ ] The package is novel and will be of interest to the broad readership of the journal.
- [ ] The manuscript describing the package is no longer than 3000 words.
- [ ] You intend to archive the code for the package in a long-term repository which meets the requirements of the journal (see [MEE's Policy on Publishing Code](http://besjournals.onlinelibrary.wiley.com/hub/journal/10.1111/(ISSN)2041-210X/journal-resources/policy-on-publishing-code.html))
- (*Scope: Do consider MEE's [Aims and Scope](http://besjournals.onlinelibrary.wiley.com/hub/journal/10.1111/(ISSN)2041-210X/aims-and-scope/read-full-aims-and-scope.html) for your manuscript. We make no guarantee that your manuscript will be within MEE scope.*)
- (*Although not required, we strongly recommend having a full manuscript prepared when you submit here.*)
- (*Please do not submit your package separately to Methods in Ecology and Evolution*)

</details>

## Code of conduct

- [x] I agree to abide by [rOpenSci's Code of Conduct](https://ropensci.org/code-of-conduct/) during the review process and in maintaining my package should it be accepted.

**Comment by maelle** [comment]
@ropensci-review-bot check package

**Comment by maelle** [comment]
@ropensci-review-bot check package

**Comment by maelle** [comment]
@micha-silver thanks for your submission! We'll fix this glitch in our system next week.

**Comment by micha-silver** [comment]
Hi @maelle : It's good to hear from you. I'll look forward to getting the review underway next week.

**Comment by noamross** [comment]
Thanks @micha-silver for your submission.  As Maëlle has noted our automatic package checker has a glitch, one that we think is a GitLab-related edge case. (We have fewer - totally fine! - submissions from GitLab so the system has had less stress-testing against it).  

In the meantime, I've determined that your package is in-scope.  From an initial manual check of the repository I note a few important things that that will be required before proceeding to assign an editor and reviewers.  It may be worth moving ahead on these while we get the automatic checks going:
 - We require that all packages have at least on vignette (https://devguide.ropensci.org/building.html#documentation)
 - We also require a CONTRIBUTING.md or similar file to provide guidance to potential package reviewers (https://devguide.ropensci.org/collaboration.html?q=contributing.md#contributing-guide)
 - We require testing against the current (release), last (oldrel), and development (devel) versions of R, as well as across linux, macos, and windows platforms (usually just on release).  We also require coverage reporting of testing.  Based on your [.gitlab-ci.yml file](https://gitlab.com/rsl-bidr/roptram/-/blob/main/.gitlab-ci.yml), I think you are only testing on `release`. (https://devguide.ropensci.org/ci.html)

**Comment by maelle** [comment]
@ropensci-review-bot check package

**Comment by maelle** [comment]
@ropensci-review-bot check package

**Comment by micha-silver** [comment]
Thanks: 
>     * We require that all packages have at least on vignette (https://devguide.ropensci.org/building.html#documentation)
I've written and added an appropriate vignette.

> 
>     * We also require a CONTRIBUTING.md or similar file to provide guidance to potential package reviewers (https://devguide.ropensci.org/collaboration.html?q=contributing.md#contributing-guide)
Added CONTRIBUTING.md (in `docs` subdir)

> 
>     * We require testing against the current (release), last (oldrel), and development (devel) versions of R, as well as across linux, macos, and windows platforms (usually just on release).  We also require coverage reporting of testing.  Based on your [.gitlab-ci.yml file](https://gitlab.com/rsl-bidr/roptram/-/blob/main/.gitlab-ci.yml), I think you are only testing on `release`. (https://devguide.ropensci.org/ci.html)

Yes, I have only one CI so far. I'll start preparing tests for additional releases and OS.

**Comment by noamross** [comment]
Thanks, @micha-silver. Now that we have automated checks, I also see that the repo doesn't show the test coverage (though you may be measuring this locally as I see you have a .covrignore file). Since reviewers may be less familiar with navigation in GitLab, please put badges in your readme to point us to the CI build logs and coverage outputs once those are set up.

**Comment by noamross** [comment]
This is a common challenge we see.  There are several complementary strategies one can take:

 - Write _mock_ tests that test that the functions work with pre-cached data or outputs: https://books.ropensci.org/http-testing/mocking.html
- Cache data required for testing in the package, or in another public location from which fetching is faster or more accessible. (Such as attached to the GitHub repository as a release, or in a Zenodo record).  Pre-process if needed or possible to reduce size.
 - Break up the functions into components that require downloading and those that don't, and write separate tests to cover the latter.
 - Make tests conditional so they don't need to be run all the time, such as making them conditional on setting an environment variable `ROPTRAM_ALL_TESTS=1`.  This would normally be turned off on CI and thus lower coverage would be reported, but that's OK. The test-running procedure should be explained in CONTRIBUTING.md and reviewers will be able to consider it.

**Comment by micha-silver** [comment]
@noamross : I'm considering migrating the rOPTRAM repo from gitlab to github. I think I'll be able to configure the CI tests more easily on github (lots more examples...). And since most ROpenSci projects are on github, the review process might also be smoother. Shall I go ahead with this? Anything I need to be aware of??

Regarding your explanation above about testing functions with large downloads, I'd like to go with the 4th option, setting a condition, and skipping those tests (with explanation in CONTRIBUTING). 

Thanks, Micha

**Comment by noamross** [comment]
@micha-silver No problem for us if you move to GitHub, though we are happy to do the process either way.  We might have to make a tweak so the bot knows that the repository is in a new location, so let us know when you switch.

**Comment by micha-silver** [comment]
Hi:
I've made progress with the `rOPTRAM` package:

1. Added examples to all functions
2. CI tests are now integrated.
3. `covr::package_coverage()` now shows:                 
```
r$> covr::package_coverage()
rOPTRAM Coverage: 75.49%
R/optram_acquire_s2.R: 17.86%
R/utilities.R: 85.29%
R/optram_wetdry_coefficients.R: 95.10%
R/optram_ndvi_str.R: 95.12%
R/optram_calculate_str.R: 95.45%
R/optram_soilmoisture.R: 100.00%
```
(**TODO:**  Still needs to be improved. I've added one addition testthat function to test downloading in the `optram_acquire_s2.R` function by limiting the timespan so that only one sentinel tile is downloaded. But that function still has low coverage. )


5. I run tests on 5 platforms: Windows release, Windows oldrel, Windows devel, MacOS, and Ubuntu. The `gitlab-ci.yml` performs these tests by sending to rhub. Here's the summary currently:

```
                  platform errors warnings notes
       ubuntu-gcc-devel      0        0     0
 windows-x86_64-release      0        0     0
  windows-x86_64-oldrel      0        0     0
   windows-x86_64-devel      0        0     0
          macos_release      0        0    11

[1] "2023-11-26 11:21:16.729992  - Check completed"
```
Here are links to the results:
for [Ubuntu: ](https://builder.r-hub.io/status/rOPTRAM_0.0.1.000.tar.gz-ca3a59ee930546f09aad13ec3adadd32)
for [Win release: release](https://builder.r-hub.io/status/rOPTRAM_0.0.1.000.tar.gz-99b307e19929472a847598acd1adade0)
for [Win devel: ](https://builder.r-hub.io/status/rOPTRAM_0.0.1.000.tar.gz-8cb1ebeb377b4b2387ce2f75f779e0ad)
for [Win oldrel: ](https://builder.r-hub.io/status/rOPTRAM_0.0.1.000.tar.gz-62efd80c37184268ab0dd9b7afb34c76)
for Mac [macbuilds:](https://mac.r-project.org/macbuilder/results/1700997075-a834c7066836f9f0/)

Not sure how long these artifacts are kept on rhub. But the full artifact on gitlab is available:
[artifacts on gitlab](https://gitlab.com/rsl-bidr/roptram/-/artifacts)

What's next?
Best regards, Micha

**Comment by micha-silver** [comment]
Following up:
On my local machine I get good coverage:


```
r$> cvr <- covr::package_coverage()

r$> cvr
rOPTRAM Coverage: 92.70%
R/utilities.R: 85.29%
R/optram_acquire_s2.R: 90.59%
R/optram_wetdry_coefficients.R: 95.10%
R/optram_ndvi_str.R: 95.12%
R/optram_calculate_str.R: 95.45%
R/optram_soilmoisture.R: 100.00%
```

The explanation: The `optram_acquire_s2()` function relies on Google Cloud CLI to download Sentinel images. I have that SDK installed locally, but I have **not seen any way** to install non-interactively, on a rocker image. There seems to be no work around to get access authentication to Google services, and the `gsutil` utility setup in a CI machine. I tried to copy my Google access token into a CI variable, and use that, but there is no way to get `gsutil` setup automatically to recognize a username and token. So when I run `covr` in `gitlab-ci` it skips the whole download procedure, and shows lower coverage.

**Comment by noamross** [comment]
Hi @micha-silver, thanks for moving a ahead on this. It's good you got to 75% coverage without the `gsutil` dependency!  I note that https://cloud.google.com/sdk/docs/downloads-interactive#silent to have some guidance for installing on CI services. I'd make good-faith effort with that, but if it is not possible a full description in the install section of the README as well as CONTRIBUTING for those running tests will be sufficient.  I note this is one thing that we might ask reviewers to comment on in the next stage - would another dependency be a simpler way to fetch the images?

**Comment by micha-silver** [comment]
@noamross I appreciate your encouraging words. 

Two comments 

1. The link you suggested to an automated install of the Gcloud CLI works fine. The problem is that after installing you still need to be authenticated. And that, AFAIK, requires an interactive step: Google forces you to go to a certain website, put in credentials and get the OAuth token on your machine. I tried to put a valid token into a gitlab CI variable, then somehow use that to manually recreate the local auth directory (in a docker image) that gcloud requires - no success here :-(.  Even if we could find some work around, it would surely be very fragile. 
2. As to a simpler way to fetch the imagery:  Indeed Copernicus now (since Oct 2023) offers several new options for downloading - many more than when I began this package. We have recruited another contributor who will help me to weed thru these various options, and implement the best, smoothest ones. If you agree, I'd like to present the package in its current state for review, then add the additional image acquisition capabilities in a future version.

**Comment by noamross** [comment]
It should be fine to present the package in its current state. Let me re-run checks.

**Comment by noamross** [comment]
@ropensci-review-bot check package

**Comment by micha-silver** [comment]
I have improved the coverage of `optram_acquire_s2()` by moving the actual download code to a separate file, and entering that new file into .covrignore.

CI tests are, if I understand correctly, implemented in the `.gitlab-ci.yml` file. In that file I put a rule to run the pipeline manually (to avoid firing off a check at every minor commit). Is that sufficient to fill the requirement for "✖️ Package has no continuous integration checks."?

I also fixed the `lintr` comments.

**Comment by noamross** [comment]
Thanks, for the CI the reason for failure is that we don't yet have a standard way to detect it on GitLab.  We'll work on that in the future, both detection and some guidance on how to signal it for non-GitHub users. In the meantime, I'm assigning an editor to get this going!

**Comment by noamross** [comment]
@ropensci-review-bot assign @adamhsparks as editor

**Comment by adamhsparks** [comment]
Hi @micha-silver, I'll start looking for reviewers, but in the meantime I've found a few issues that should be corrected before they start their reviews.

## Documentation

* You have wrapped "sen2r" in "{}", the proper way to refer to other packages in documentation is either, `\pkg{}` or if it's on CRAN, `\CRANpkg{}`. The latter will link to the package docs.
* Function titles should be in title case, not sentence case.

## Package Overall

* Check the spelling in the whole package. I typically will use `spelling::update_wordlist()` to generate a list that will be used to check against but first I check carefully against the list of words it thinks are misspelled and correct any first and then rerun it. To do this you should also set the spelling in your DESCRIPTION file, `Spelling: en-US`.
	* Check the spelling in your README.Rmd and vignette as well.
	* In some docs, "boolean" should be spelled as "Boolean". 
* Avoid using `T` and `F`, spell out the whole word, please.
* In the Description field, CRAN usually just asks for the authors, year and DOI, _e.g._, https://github.com/gabrielblain/PowerSDI/blob/e222da9f53c837a26700fee1feaa037c82fce4c2/DESCRIPTION#L33
* If your functions in Utilities.R don't need to be exported and user-facing, you can use `@noRd` and `@keywords Internal` so that they aren't exported and documented. Glancing over them, I'm unsure why a user would call them, but I may just not be familiar enough with the workflow yet.
* You have some files in the top level of the package structure that are not usual, _e.g._, "paper.md", "bibliography.bib", these appear in .Rbuildignore, but might be better in "inst" to follow common R package structuring.
* I note that you have a vignette Rmd and an Rmd.orig, but no script file to compile the Rmd, this seems unusual to not have a way to compile the resulting Rmd, you might consider adding it to the "vignettes" directory, _e.g._, https://github.com/gabrielblain/PowerSDI/tree/master/vignettes

**Comment by micha-silver** [comment]
Hello @adamhsparks 
Thanks for your helpful comments. I've addressed most:

1. Corrected formatting for imported packages in help files;
2. Formatted all function titles in Title Case;
3. Spell check done;
4. Corrected 'T' to 'TRUE';
5. Removed extra bibliography details from DESCRIPTION;
6. As for _paper.md_ and _bibliography.bib_, I intend to submit a short paper with the package. Is this not the proper place to have the paper in the same repo as the code?
7. Regarding the vignette, I guess I don't yet understand how this should work :-( .  When I build a source tarball, then install from that, as suggested in [r-bloggers](https://www.r-bloggers.com/2020/05/how-to-add-a-vignette-to-a-package-in-rstudio/) then ` browseVignettes(package = "rOPTRAM") ` works as expected. Am I missing something?

**Comment by micha-silver** [comment]
@adamhsparks :

OK, that's clear now. I've moved the paper into /inst, and added the precompile script as you suggested.

One glitch, maybe you have some idea: I want a tmap image in the vignette (the final code chunk). However when I knit the rOPTRAM.Rmd.orig, I get this error at the end, inside the final compiled rOPTRAM.Rmd:

`#> Error in path.expand(path): invalid 'path' argument`

and there's no image in the output HTML, just the same error message.

- Tried both in and RStudio session, and an R command prompt
- Tried also with leaflet, and I get the same error.
- Running the code directly (not thru knitr) works as expected: image appears in the viewer.

Any ideas?

Thanks, Micha

**Comment by adamhsparks** [comment]
Hi @micha-silver,
I'm trying to troubleshoot this for you but I'm having difficulties getting everything set up. I think one thing that will be necessary for users of this package will be to have very clear instructions on how to set up the computing environment to use this package as the normal `install.packages("rOPTRAM")` won't work here.

I can get the vignette to properly knit using this code, which avoids using `setwd()` so it is portable. This also fixes the output directory, since I don't have a `"/home/micha/EO_Data/Israel/Migda_full"` directory to knit into, again, portability.

```r
# vignette that depends on Internet access need to be pre-compiled and takes a
# while to run
library("knitr")
knit("vignettes/rOPTRAM.Rmd.orig", "vignettes/rOPTRAM.Rmd")

# remove file path such that vignettes will build with figures
replace <- readLines("vignettes/rOPTRAM.Rmd")
replace <- gsub("<img src=\"vignettes/", "<img src=\"", replace)
fileConn <- file("vignettes/rOPTRAM.Rmd")
writeLines(replace, fileConn)
close(fileConn)

# build vignette
library("devtools")
build_vignettes()

# move resource files to /docs
resources <-
  list.files("vignettes/", pattern = ".png$", full.names = TRUE)
file.copy(from = resources,
          to = "docs",
          overwrite =  TRUE)
```

It knits, but when I tried to step through the examples, the examples in the vignette don't work for me due to the use of hardcoded directories, once again. Here I used, `tempdir()` to fix that and I'm unable to configure my "Google Cloud" environment properly for some reason, so I cannot generate the whole vignette as it should appear.

Maybe this helps you get it to knit properly and will be useful for the reviewers to suggest ways that the documentation can be improved on how to set up the computing environment.

Also, I noted that,
* you don't have {tmap} in the Suggests, it is used in the vignette,
* your vignette header does not have `%\VignetteDepends{tmap}`, and
* your vignette header has` %\VignetteEngine{knitr::rmarkdown}` not  `%\VignetteEngine{knitr::rmarkdown_notangle}`

**Comment by adamhsparks** [comment]
### Editor checks:

- [x] **Documentation**: The package has sufficient documentation available online (README, pkgdown docs) to allow for an assessment of functionality and scope without installing the package. In particular,
    - [x] Is the case for the package well made?
    - [x] Is the reference index page clear (grouped by topic if necessary)?
    - [x] Are vignettes readable, sufficiently detailed and not just perfunctory?
- [x] **Fit**: The package meets criteria for [fit](https://devguide.ropensci.org/policies.html#package-categories) and [overlap](https://devguide.ropensci.org/policies.html#overlap).
- [x] **Installation instructions:** Are installation instructions clear enough for human users?
- [x] **Tests**: If the package has some interactivity / HTTP / plot production etc. are the tests using [state-of-the-art tooling](https://devguide.ropensci.org/building.html#testing)?
- [x] **Contributing information**: Is the documentation for contribution clear enough e.g. tokens for tests, playgrounds?
- [x] **License:** The package has a CRAN or OSI accepted license.
- [x] **Project management**: Are the issue and PR trackers in a good shape, e.g. are there outstanding bugs, is it clear when feature requests are meant to be tackled?
---

#### Editor comments

Everything checks out with the normal editor checks. I'll start looking for reviewers now.

---

**Comment by micha-silver** [comment]
My last commit:

- Adds {tmap} and {here} in Suggests:
- Avoids hard coded paths in the vignette by using the `here()` function. (Thanks for saving my computer from being set on fire :-)) by Jenny )
- Corrects another typo that you found.
- Adds a few lines in the vignette, in parallel with the README, showing a link to the Google Cloud installation instructions. Indeed getting Google Cloud CLI working on various platforms is ... challenging. As I mentioned, we hope in the coming months to implement some of the newer download options now offered by the ESA in their "Copernicus Dataspace" Platform. Until then users will need to run the Google Cloud installation, then perform the **interactive** `gcloud init` to get Google's token installed on their computer. 
- Still no map plot in the precompiled Rmd. :-(  I'm unsure what to do about this. (Running `knit` on the rOPTRAM.Rmd.orig file directly to HTML output does produce the inline map image). 

Best regards,

**Comment by adamhsparks** [comment]
@micha-silver, I'll have a look tomorrow if I have some time. Thanks for all the effort thus far!

**Comment by adamhsparks** [comment]
Hi @micha-silver, I would advise against your examples writing to "./inst" in the vignette as you do when saving the data. There is also no need to use {here} in your package (aside from the pre-compile function, which isn't really part of the R package so {here} won't need to be in the DESCRIPTION file anywhere) for these examples, just use `tempdir()` for your vignette examples as I previously suggested. You're vignette also indicates in the text that the downloads are saved to `tempdir()` whereas you're putting them in "./inst".

Also, there is no need to install {rOPTRAM} in the vignette example. One could assume that the user browsing the documents locally has installed it locally so that they can view the docs locally.

Unfortunately I can't really troubleshoot the map for you since I can't run the functions in the vignette due to difficulties with the Google tools that you require, sorry. Perhaps the reviewers will be able to assist with that.

**Comment by micha-silver** [comment]
@adamhsparks  Thanks for your patience with this question. 

I'm hesitating to use `tempdir()` only because the download for each compile of the vignette is around 50GB, and takes around 1.5 hrs. (necessary to get a reasonable/realistic trapezoid from the model). If I define a local, permanent directory for my precompile script, then the package "sees" the files already downloaded from a previous compile, and completes the vignette in a few minutes.  If I use `tempdir()` then obviously each minor change to the vignette will require a long wait to see the results.

In any case, let's wait for reviewers, and then I'll revisit this issue.

**Comment by micha-silver** [comment]
@adamhsparks Thanks for your help in enlisting reviewers. Looking forward to working with @harryeslick

**Comment by adamhsparks** [comment]
@ropensci-review-bot add @obrl-soil as reviewer

**Comment by adamhsparks** [comment]
@obrl-soil, as already agreed, please take some extra time with your field work and the holidays coming up. I’ll follow up mid-late January if I don’t hear back from you by then.

**Comment by harryeslick** [comment]
Hi @micha-silver ,
I'm looking forward to working with you on this review. As mentioned by adamhsparks I am on leave until mid-Jan so won't have much time to meaningfully contribute before then. 

I've just been having a superficial look and thought I would share some thoughts:
- the prerequisites section was not clear to me. I prefer to explicitly name things rather than referring to them with this or that. Eg. 

>This package uses the R package for acquiring Sentinel-2 imagery. To install that package...

Could be made clear using curly bracket convention to show R packages eg. 

{rOPTRAM} uses {sen2r} for acquiring Sentinel-2 imagery. To install {sen2r}...


- It appears that {sen2r} has recently been retired. ([sen2r will be retired from CRAN by the end of 2023](https://github.com/ranghetti/sen2r)). I don't think this will affect the current review, but suggests an update maybe required fairly soon.

**Comment by micha-silver** [comment]
Hi @harryeslick :

Thanks for the early comments. I'll have a look at replacing the references to external packages, as you suggested. 

Regarding {sen2r}, indeed the maintainer has gone to other things, and what's worse, part of his earlier implementation (accessing the Copernicus API) no longer works since the API has been changed. We are now actively working on alternatives. This will probably include the new {CDSE} package, now on CRAN, that offers access thru the new Copernicus API.
I'd like to go ahead with the review process looking at the package as is, currently. If it's accepted to ROpenSci, then we'll move on to a second version once we have implemented and tested the image acquisition changes.

Looking forward to further responses ... next year. Happy holidays.

Regards, Micha

**Comment by adamhsparks** [comment]
I think that it's necessary to remove any dependencies on a package that is no longer supported, _i.e._, {sen2r}. Please make the necessary changes to ensure that {rOPTRAM} will work as anticipated by users before we accept it into rOpenSci so that the reviewers may review the package as it will be used.

**Comment by micha-silver** [comment]
OK, so I understand the review is on "freeze" until I remove the deprecated {sen2r}, and implement the new API. 
We're now working on the replacements for {sen2r}, so the complete package should be ready "soon...".

**Comment by adamhsparks** [comment]
@ropensci-review-bot put on hold

**Comment by adamhsparks** [comment]
@harryeslick and @obrl-soil, I've placed a "Holding" label on this review; please hold off on your reviews until the appropriate changes have been made to the package. If you are unable to complete your reviews at a later date, just ping me and I'll get it sorted.

**Comment by micha-silver** [comment]
Hello @adamhsparks 
I'm happy to inform that we've completed removal of the deprecated {sen2r} package, and replaced it with the new {CDSE} package for access to Sentinel imagery. We also added additional vignettes, as well as two new curve fitting options to create the OPTRAM 'trapezoid'. 
Today I ran the CI pipeline, and all tests passed on the gitlab installation (ubuntu release).  Coverage is at 87.9%.  I also request tests, from the CI pipeline  using rhub on Windows and Mac, and these are failing since a few tests require login credentials. (I don't yet know how to send the creds to rhub. I'll check how to use environment variables...)

On Ubuntu release:
```
── R CMD check results ────────────────────────────────── rOPTRAM 0.1.0.000 
Duration: 1m 35.8s
❯ checking installed package size ... NOTE
    installed size is  8.1Mb
    sub-directories of 1Mb or more:
      doc       2.2Mb
      extdata   5.6Mb
0 errors ✔ | 0 warnings ✔ | 1 note ✖
rOPTRAM Coverage: 87.96%
```

Other platforms:
```
ℹ Check
  <https://mac.r-project.org/macbuilder/results/1709624544-1d521aff36a481d0/>
  the results in 5-10 mins (~07:52 AM).
               platform errors warnings notes
 windows-x86_64-release      1        0     1
  windows-x86_64-oldrel      0        0     0
   windows-x86_64-devel      1        0     1
          macos_release      6        4    18
[1] "2024-03-05 07:52:26.288169  - Check completed"
[1] "Elapsed time for checking:  24.1 minutes"
Error: Some checks with ERROR, or WARNING.
Execution halted
```

Please consider releasing the hold, and approaching reviewers to begin their review of the package.
Thanks,

**Comment by adamhsparks** [comment]
@ropensci-review-bot remove hold

**Comment by adamhsparks** [comment]
@ropensci-review-bot help

**Comment by micha-silver** [comment]
Update: 
I added a `skip_if_not` to avoid the test for the credentials when the creds file is not available. So now the Windows builds go thru cleanly. But MacOS still not :-(

**Comment by adamhsparks** [comment]
@harryeslick and @obrl-soil, sorry for the delay. I'll set the due date to three weeks from now. If you need more time, feel free to take it, please just let us know you need more time.

**Comment by adamhsparks** [comment]
@ropensci-review-bot set due date for @harryeslick  to 2024-03-25

**Comment by adamhsparks** [comment]
@ropensci-review-bot set due date for @harryeslick to 2024-03-25

**Comment by adamhsparks** [comment]
@ropensci-review-bot set due date for @obrl-soil to 2024-03-25

**Comment by micha-silver** [comment]
Hi @adamhsparks 
Just a quick update: we've fixed the final two (minor) issues with the MacOS build. Now I'm getting no errors and no warnings on all 5 platforms.

**Comment by micha-silver** [comment]
Hi @adamhsparks 
No response yet from either of the reviewers. Is it time for a gentle reminder?

**Comment by obrl-soil** [comment]
Hey - sorry but the rewrite pushed this from a quiet time for me to a very busy one! I can get to this over the long weekend.

**Comment by harryeslick** [comment]
Hi @micha-silver ,
Im working my way through. 
I've logged an issue for you in GitLab: `rsl-bidr/roptram#1`

**Comment by micha-silver** [comment]
@harryeslick 
Thanks for finding time for the review.  
This issue is fixed in the gitlab repo:  [99c04bff](https://github.com/rsl-bidr/roptram/-/commit/99c04bff9ce0957b9a6b748cb68f2ff4e251d367). 
I improved the check_swir_band function, and also changed the example.

**Comment by harryeslick** [comment]
@micha-silver , here is my review, 
For the most part it all works quite easily, so well done! my comments as all  minor things. 
 
I still need to read up on the `Packaging guidelines` so Ive not quite completed it yet, so some jobs for me to follow up on as well. 

## Package Review

*Please check off boxes as applicable, and elaborate in comments below.  Your review is not limited to these topics, as described in the reviewer guide*

- **Briefly describe any working relationship you have (had) with the package authors.**
	- nil
- [x] As the reviewer I confirm that there are no [conflicts of interest](https://devguide.ropensci.org/policies.html#coi) for me to review this work (if you are unsure whether you are in conflict, please speak to your editor _before_ starting your review).

#### Documentation

The package includes all the following forms of documentation:

- [x] **A statement of need:** clearly stating problems the software is designed to solve and its target audience in README
- [x] **Installation instructions:** for the development version of package and any non-standard dependencies in README

some minor suggestions:
##### OAuth link missing
Running the main wrapper function `optram()` in the example states that 
> Registration on Copernicus DataSpace was done in advance, and the OAuth credentials (_clientid_, and _secret_) have been saved to the user’s home directory.

I would  i like to have a link here to the vignette where setting up OAuth is demonstrated. 

##### OAuth example suggestion
This is a bit of a stylistic preference, and keep in mind I don't fully understand the workflow, so I could be totally wrong here...
`roptram/docs/articles/rOPTRAM_acquiring.html#step-4-saving-credentials`

Saving credentials section seems a little long winded and unnecessary. I think it could probably be replaced with just 
```
store_cdse_credentials("my_clientID", "My_secret")
```

the convenience of having a ` save_creds = TRUE,` seems to require more time to explain that it saves in the actual workflow.  but I could be wrong. 


- [x] **Vignette(s):** demonstrating major functionality that runs successfully locally

- [ ] **Function Documentation:** for all exported functions
### Missing function in docs
main page references function `optram_soil_moisture()` 
    `roptram/docs/index.html#optram_soil_moisture`

Is this meant to be `optram_calculate_soil_moisture()`?

ALSO: `optram_calculate_soil_moisture` fails with default `trapezoid_method`
`trapezoid_method` is called (Line92) before  `arg.match(trapezoid_method)` (Line108)
* related to https://gitlab.com/rsl-bidr/roptram/-/issues/1

- [ ] **Examples:** (that run successfully locally) for all exported functions

I found a few minor bugs here:

`acquire_scihub(aoi, from_date, to_date, veg_index = "SAVI")`
returns error
> SWIR band must be either 11 or 12
* related to https://gitlab.com/rsl-bidr/roptram/-/issues/1

`optram_acquire_s2`
1. comma missing after parameter `remote = "scihub"`
2. `SWIR_Band` default fails: 
    * related to https://gitlab.com/rsl-bidr/roptram/-/issues/1

`optram_calculate_soil_moisture`
1. VI_dir  "SAVI" does not exist. Should this be "NDVI"?
2. data_dir  `system.file` call missing package parameter. 
   should be  `data_dir  <- system.file("extdata", package = "rOPTRAM")`

`optram_calculate_str`
returns `NULL`
data_dir  `system.file` call missing package parameter. 
should be: `BOA_dir <- system.file("extdata", "BOA", package = "rOPTRAM")`

`optram_ndvi_str`
returns `NULL`
VI_list  `system.file` call missing package parameter.  Also using `SAVI` instead of `NDVI`
STR_list  `system.file` call missing package parameter. 

`optram_wetdry_coefficients`
Please check example here. not sure if the plot works...

`plot_vi_str_cloud`
Please check example here. not sure if the plot works...

`retrieve_cdse_credentials`
No example provided, but not sure if one is required. 
Does this function need to be exported?

`store_cdse_credentials`
No example provided


- [x] **Community guidelines:** including contribution guidelines in the README or CONTRIBUTING, and DESCRIPTION with `URL`, `BugReports` and `Maintainer` (which may be autogenerated via `Authors@R`).

#### Functionality

- [x] **Installation:** Installation succeeds as documented.
- [x] **Functionality:** Any functional claims of the software have been confirmed.
- [x] **Performance:** Any performance claims of the software have been confirmed.
- [x] **Automated tests:** Unit tests cover essential functions of the package and a reasonable range of inputs and conditions. All tests pass on the local machine.
- [ ] **Packaging guidelines**: The package conforms to the rOpenSci packaging guidelines.
    

Estimated hours spent reviewing: 6 

- [x] Should the author(s) deem it appropriate, I agree to be acknowledged as a package reviewer ("rev" role) in the package DESCRIPTION file.

---

### Other related issues logged at GitLab
#### `rOPTRAM::optram` default argument `SWIR_band` is invalid

issue related to default parameter values logged as issue here: https://gitlab.com/rsl-bidr/roptram/-/issues/1

#### CDSE credentials saved and retrieved from different file paths
credentials bug logged :  https://gitlab.com/rsl-bidr/roptram/-/issues/2

**Comment by micha-silver** [comment]
@harryeslick 
With the latest commit we have (I think) addressed all the suggestions that you raised. Any further comments are welcome!
(How can I get your email addr to add you as 'rev' to the DESCRIPTION file?)

**Comment by obrl-soil** [comment]
Hey, sorry this took so long! Not sure I caught everything but I think I've covered off on the main issues. Tl;dr, good job overall but I think the function design is trying to do a little too much for the end user in places.

## Package Review

- **Briefly describe any working relationship you have (had) with the package authors.**
- [x] As the reviewer I confirm that there are no [conflicts of interest](https://devguide.ropensci.org/policies.html#coi) for me to review this work (if you are unsure whether you are in conflict, please speak to your editor _before_ starting your review).

#### Documentation

The package includes all the following forms of documentation:

- [x] **A statement of need:** clearly stating problems the software is designed to solve and its target audience in README
- [x] **Installation instructions:** for the development version of package and any non-standard dependencies in README
- [x] **Vignette(s):** demonstrating major functionality that runs successfully locally
 
  - see below
  
- [x] **Function Documentation:** for all exported functions

  - see below

- [ ] **Examples:** (that run successfully locally) for all exported functions

  - see below

- [x] **Community guidelines:** including contribution guidelines in the README or CONTRIBUTING, and DESCRIPTION with `URL`, `BugReports` and `Maintainer` (which may be autogenerated via `Authors@R`).

#### Functionality

- [x] **Installation:** Installation succeeds as documented.
- [x] **Functionality:** Any functional claims of the software have been confirmed.
- [x] **Performance:** Any performance claims of the software have been confirmed.
- [ ] **Automated tests:** Unit tests cover essential functions of the package and a reasonable range of inputs and conditions. All tests pass on the local machine.
- [ ] **Packaging guidelines**: The package conforms to the rOpenSci packaging guidelines.

  - Do you need a [CITATION file](https://devguide.ropensci.org/pkg_building.html#citation-file)?
  - The [README guideline](https://devguide.ropensci.org/pkg_building.html#readme) suggests you may need a few more badges - code coverage, etc
  - Roxygen formatting could be improved in several places (see below)

Estimated hours spent reviewing: 8

- [x] Should the author(s) deem it appropriate, I agree to be acknowledged as a package reviewer ("rev" role) in the package DESCRIPTION file.

---

### Review Comments

 - testing against version installed 2024-04-13 on Windows 11, R 4.3.3, RStudio "Ocean Storm" Release (4da58325, 2024-01-28)
 - used QGIS desktop 3.34.2 to check outputs also
 
#### General

- I don't think that `plot_vi_str_cloud()` should be calling `ggsave()`; let the user do that (you can demonstrate in an example) or at the very least make it optional. Depending on what the user wants to do with the graphic they'll likely want more control over dimensions, DPI, etc. Also, the way its written this function never sends a plot to the graphics device - the outputs can't be viewed in RStudio at all. This is, I think, not generally expected when `plot()` is called.
- I don't think your functions should be reading spatial data in for the users, and I'm specifically focusing on the `aoi_file` parameter here. It creates a lot of room for weird errors which you'll then be pestered about. Instead, I think it would be far safer to require the user to read in a file using `sf` and supply the resultant object to `aoi_file`. This then gives you the ability to write unit tests and error catches to cover e.g. correct spatial type supplied, data projected in an acceptable CRS, non-insane extents, etc. You can't do any of that when all you have to work with is a file path string.
 
#### Vignettes

  - `remotes::install_gitlab()` does not build vignettes by default. You may wish to specify installation with `build_vignettes = TRUE` in your README to prevent confusion.
  - As mentioned below, "Acquiring Sentinel-2 Imagery: Step 3" will need an update to match the current state of the website.
  - The `knitr::include_graphics()` calls in Vignette "rOPTRAM: Three Trapezoid Fitting Methods" are failing with a 'can't find file' error - might be related to installing from gitlab though.
 
#### Documentation

- Documentation is good (I especially like how you've specified the data types for each arg) but would benefit from some finessing, namely:
  
  - Use \code{} rather than backticks for code-formatting
  - Use internal cross-links to other functions like \code{\link[rOPTRAM:optram_aquire_s2]{optram_aquire_s2()}}}
  - Use \url{url} or \href{url}{alt-text} for external URL links
  - Use \doi{} for DOI links
  - Use roxygen list formatting in the more complex 'notes' fields. 
  - math formatting is available for formulae with \eqn{} or \deqn{}
    
see https://roxygen2.r-lib.org/articles/formatting.html .
  
For specific functions,

**`aquire_scihub()`**

- Most people will be thinking of another website entirely when they see the word scihub 😛 - is there an appropriate alternate name? 
- The instructions in the `note` for `aquire_scihub()` seem to be out of date; I think the website has been updated. I had to do the following:
  - Register and log in.
  - On https://dataspace.copernicus.eu/, click on 'my account' in the top right
  - When the account console loads, click on 'Sentinel Hub' in the 'Dashboards' panel
  - Click on user settings in the bottom left of the screen
  - Follow the prompts to create an oauth client as instructed
- So this took quite a bit of extra clicking around to figure out. Also, on creating the client, there is a checkbox for 'Client will be used by a single-page application'. I assume I don't need to tick this?
- Once the client was created, I tried to follow the next part of the instructions: "the user must save her secret ...etc" Unfortunately this function is not exported so the user can't run it directly - need to reconsider including this instruction.

**`exponential_coefficients()`**

- missing full stop at end of first sentence in description

**`optram_landsat()`**

- typo in description, 'Use this function to prepare**s**...'

**`optram_wetdry_coefficients()`**
  
- The Note mentions that "The vegetation index column is named "VI" though it can represent several vegetation indices, such as SAVI, or MSAVI." Would it not be preferable to name it something more generic like 'veg_index'?
- Speaking of column naming conventions, https://emilyriederer.netlify.app/post/column-name-contracts/ provides some ideas for consistent and informative column names. This is a valuable approach for anyone outputting to multiple on-disk formats, as you are (TIF, CSV, etc)

#### Examples

- examples for non-exported functions need to use `rOPTRAM:::` syntax to work correctly
- examples using the `landsat_dir` parameter should all explain what's going on there, currently only `optram_landsat()` does

**`aquire_openeo()`**
  
- Example fails as SWIR band not supplied - tested with 11 and then 12. 
- Example prompts in console to copy/paste an ID code that is subsequently not asked for. Not sure if that's an openeo problem? Looks like e.g. " → Copy CTGB-UGFU and paste when requested by the browser"
- Important! Example code should assign the output to an object, e.g. `acq <- aquire_openeo()...` otherwise the list of files just gets reported to the console and is otherwise impossible to retrieve without re-running the code.
- The retrieval process throws an error: "[INFO] 'custom_processes' not loaded: ModuleNotFoundError("No module named 'custom_processes'")." The extraction and processing seemed to run to completion despite this, but I can't be sure as I don't know what success should look like for this function. I have folders in my temp dir for BOA, STR and SAVI containing Geotiffs for 14 days within the range requested but can't tell if they've actually been processed correctly.
- This function ran long for unknown reasons on a subsequent test - 'job still running' just kept printing and I had to kill the session to continue reviewing.
  
**`aquire_scihub()`**

- As with `aquire_openeo()` the SWIR band needs to be specified or checked differently.
- Although of course you can't include an actual clientid and secret, the example code should maybe look more like:
    
```
acq_sh <- acquire_scihub(aoi, from_date, to_date,
            veg_index = "SAVI",
            clientid = 'paste id here',
            secret = 'paste secret here',
            SWIR_band = 11)
```	
  to prompt the user directly.
  
- This function ran much more quickly than `acquire_openeo()` and didn't report any errors to the console. I see it masks the output to the AOI polygon without being prompted too - is that a desirable/possible option for `acquire_openeo()`?

**`calculate_vi()`**

- Example as written produces an all-NaN output

**`crop_landsat_list()`**  

- Example as written can't be run (appears to lead on from an example started elsewhere). Wrapping it in 'dontrun' isn't sufficient here, it should work if the user activates it deliberately.
  
**`exponential_soil_moisture()`, `linear_soil_moisture()`, `optram_calculate_soil_moisture()`**

- final command in each example doesn't run - `data_dir = ` might need editing?
  
**`optram_acquire_s2()`**

- comma missing after `remote = "scihub"` in final command

**`optram_calculate_str()`**

- The example outputs all have pixels with value ranges of ~1-5, except for a single outlier pixel that has a value of ~58. Is that expected behaviour? Pixel was on a patch edge, so I thought I'd better check.

**`optram_ndvi_str()`**

- parameter `VI_list` returns NULL

**`optram_prepare_other_vi_str()`**

- a `print()` command referencing a nonexistent R source file should not be here. This example should be rewritten or removed.

**`polynomial_coefficients()`**

- an example should not simply read in a csv and then reference a temp dir. This example should be rewritten or removed.

**`store_cdse_credentials()`**

- uh, whose credentials are those in the example? Should they be there??

#### Tests

- In `test-aquire_s2.R`, I don't think that test 'AOI file is not spatial' is doing what it claims to. As discussed, given that your functions point to on-disk files rather than making the user read the data in first, I think the closest you could get here is running a mime-type check on the file extension. 
- More importantly, that kind of testing should be in the function itself, not in the test suite. It looks like in general quite a few of your tests are doing checks that should be in the functions themselves. The tests should really be focused more on the outputs of each function than their inputs; making sure they're as expected. They should only fail if *you* make a change to the package. 
- If you want to test for proper date formatting, you should maybe be internally converting supplied date strings to date-specific data types. I know you probably have to paste them into a string for web calls eventually, but that still gives you more options for handling and error checking.

**Comment by micha-silver** [comment]
Hi @obrl-soil 
Thanks for the detailed review! We'll start going over, point by point, and let you know when everything has been addressed.
(Can you send me your email, so that I can add you as 'rev' in the DESCRIPTION file?)

**Comment by micha-silver** [comment]
@adamhsparks , @obrl-soil , @harryeslick :
In @obrl-soil 's review, she raises an interesting point:

> "I don't think your functions should be reading spatial data in for the users, and I'm specifically focusing on the aoi_file parameter here. It creates a lot of room for weird errors which you'll then be pestered about. Instead, I think it would be far safer to require the user to read in a file using sf and supply the resultant object to aoi_file. This then gives you the ability to write unit tests and error catches to cover e.g. correct spatial type supplied, data projected in an acceptable CRS, non-insane extents, etc. You can't do any of that when all you have to work with is a file path string."

I chose to allow users to enter a file name for the area of interest, instead of requiring them to read their spatial file as an `sf` object, and passing that thru to the various functions. This choice was for two reasons: I use the file name (without file extension) later to add a title to the derived scatter plot. And second, I do not assume users would necessarily have knowledge of the {sf} package. However @obrl-soil suggests not to save the scatterplot to an image file at all, rather let the user do that by herself. So the first reason above would be canceled, leaving only the second reason: user's handling of {sf} objects by herself. 

I'd like to ask your opinions if it's reasonable to expect user's to understand the {sf} package well enough to read in their `aoi_file`, and pass an `sf` object to the OPTRAM functions. (This would certainly save some complexity in the rOPTRAM package)

Thanks,

**Comment by adamhsparks** [comment]
I agree with @obrl-soil on this. Too many edge cases. Users can be expected to have a modest understanding of {sf} to use a package that leverages it. Simplify your package. Future you will thank you. 

In the past I added “support” for operations outside the focus of my package, in time I simplified following the UNIX philosophy. Do one thing. Do it well. No one ever complained about the “loss” of functionality as it was already in R with other packages.

**Comment by adamhsparks** [comment]
I forgot to do this.

@ropensci-review-bot submit review https://github.com/ropensci/software-review/issues/612#issuecomment-2027336215 time 6

**Comment by adamhsparks** [comment]
@ropensci-review-bot submit review https://github.com/ropensci/software-review/issues/612#issuecomment-2027336215 time 6

**Comment by adamhsparks** [comment]
@ropensci-review-bot submit review https://github.com/ropensci/software-review/issues/612#issuecomment-2053591926 time 8

**Comment by micha-silver** [comment]
Hello all:
First a big thanks to the reviewers for their insightful comments. :clap: :clap:
We have gone over the list and modified as follows:

- checked the examples, all should be working (hopefully :grin: );
- went over the tests, and removed those tests that only checked input (inputs are validated within the functions). Almost all tests now check only for function output;
- adopted the suggestion to input the area of interest as an `sf` object;
- the scatterplot function does not save to a file, rather the function returns a ggplot object for further tweaking by the user;
- improved documentation where necessary;
- recompiled vignettes, with a corrected path to images.

Looking forward to any additional suggestions.

**Comment by micha-silver** [comment]
Also... I'm looking for advice on one issue:
We implement two functions for building the model from pre-downloaded Landsat, or Sentinel imagery.  Users who choose to download manually can build the module using `optram_safe()` on a directory of `*SAFE` folders, and `optram_landat()` with a directory of Landsat folders. I do not see any way to show examples, or write tests since both of these require many GB of imagery as input - way too much for a github repo. 
How should I present this?  Currently, in the examples, I sidestep the problem and put:
```
#' \dontrun{
#' aoi <- sf::st_read(system.file("extdata",
#'                               "lachish.gpkg", package = "rOPTRAM"))
#' # landsat_dir is a directory containing the original downloaded Landsat images.
#' landsat_dir <- "...enter full path here..."
#' optram_landsat(landsat_dir,  aoi,
#'                veg_index = 'SAVI',
#'                LC_output_dir = tempdir(), data_output_dir = tempdir())
#' }
```

**Comment by adamhsparks** [comment]
@ropensci-review-bot submit response https://github.com/ropensci/software-review/issues/612#issuecomment-2084811325

**Comment by adamhsparks** [comment]
> Also... I'm looking for advice on one issue: We implement two functions for building the model from pre-downloaded Landsat, or Sentinel imagery. Users who choose to download manually can build the module using `optram_safe()` on a directory of `*SAFE` folders, and `optram_landat()` with a directory of Landsat folders. I do not see any way to show examples, or write tests since both of these require many GB of imagery as input - way too much for a github repo. How should I present this? Currently, in the examples, I sidestep the problem and put:
> 
> ```
> #' \dontrun{
> #' aoi <- sf::st_read(system.file("extdata",
> #'                               "lachish.gpkg", package = "rOPTRAM"))
> #' # landsat_dir is a directory containing the original downloaded Landsat images.
> #' landsat_dir <- "...enter full path here..."
> #' optram_landsat(landsat_dir,  aoi,
> #'                veg_index = 'SAVI',
> #'                LC_output_dir = tempdir(), data_output_dir = tempdir())
> #' }
> ```

This is how I would handle the example.

Do the data requirements expand beyond the need to download the data? That is, once the data is downloaded, is the pathway for the model the same and is therefore tested with the normal tests?

**Comment by micha-silver** [comment]
> 
> This is how I would handle the example.
> 
> Do the data requirements expand beyond the need to download the data? That is, once the data is downloaded, is the pathway for the model the same and is therefore tested with the normal tests?

Yes, that's exactly what happens.

**Comment by adamhsparks** [comment]
I'd not worry about tests for this then.

**Comment by micha-silver** [comment]
@ropensci-review-bot submit response https://github.com/ropensci/software-review/issues/612#issuecomment-2084811325

**Comment by micha-silver** [comment]
I also added an initial [NEWS.md](https://gitlab.com/rsl-bidr/roptram/-/blob/main/NEWS.md?ref_type=heads)

**Comment by harryeslick** [comment]
Nothing further to add from me 👍🎉

**Comment by micha-silver** [comment]
@adamhsparks @obrl-soil 
ping...

**Comment by adamhsparks** [comment]
@micha-silver, sorry, I'm still waiting on Lauren as well.

**Comment by obrl-soil** [comment]
Hey sorry - I've no time to look over this in detail but am happy with the response as documented above.

**Comment by adamhsparks** [comment]
Thank you @harryeslick and @obrl-soil.

**Comment by adamhsparks** [comment]
@ropensci-review-bot approve roptram

**Comment by micha-silver** [comment]
@ropensci-review-bot finalize transfer of rOPTRAM
